### PR TITLE
Feature/composite interaction base

### PR DESCRIPTION
**Pull request by will-saunders-ukaea**
# Description

Please include a summary of changes, motivation and context for this PR.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] API change (breaks downstream workflows and requires major semver bump)
- [ ] Requires documentation updates

# Testing

Please describe the tests that you ran to verify your changes and provide instructions for reproducibility. Please also list any relevant details for your test configuration.

- [ ] Test Foo in /test/path/to/file_for_test_Foo.cpp
 - Description of test Foo
- [ ] Test Bar in /test/path/to/file_for_test_Bar.cpp
 - Description of test Bar

**Test Configuration**:

* OS:
* SYCL implementation:
* MPI details:
* Hardware:

# Checklist:

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any new dependencies are automatically built for users via `cmake`
- [ ] I have used understandable variable names
- [ ] I have run `clang-format` against my `*.hpp` and `*.cpp` changes
- [ ] I have run `cmake-format` against my changes to `CMakeLists.txt`
- [ ] I have run `black` against changes to `*.py`
- [ ] I have made corresponding changes to the documentation
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings

######### DELETE THIS AND BELOW #########

# How-to:

1. Format C++ source code with clang-format

```
find ./src ./include ./test -iname \*.hpp -o -iname \*.cpp | xargs clang-format -i
```

2. Format cmake files with cmake-format

```
cmake-format -i CMakeLists.txt src/nektar/CMakeLists.txt test/CMakeLists.txt
```

3. Format python with black

```
find ./python -iname \*.py | xargs black
```

### CODE REVIEW

File: src/nektar_interface/composite_interaction/composite_transport.cpp [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+                       shape_type == LibUtilities::eQuadrilateral,
+                   "unknown composite shape type");
+        geometry_composites.push_back({geom, ix});
+      }
+    }
+  }
+
+  // pack each geometry object and reuse the local_id which originally stored
+  // local neso cell index to store composite index
+  set<int> send_ranks_set;
```
Comment (JamesEdgeley) [inline src/nektar_interface/composite_interaction/composite_transport.cpp:171]:
```suggestion
  std::set<int> send_ranks_set;
```

File: test/unit/nektar_interface/test_composite_interaction.cpp [hunk 2]
```diff
@@ -0,0 +1,993 @@
...
+#include <gtest/gtest.h>
+#include <iostream>
+#include <memory>
+#include <neso_particles.hpp>
+#include <random>
+#include <set>
+#include <string>
+#include <vector>
+
+using namespace std;
```
Comment (JamesEdgeley) [inline test/unit/nektar_interface/test_composite_interaction.cpp:21]:
```suggestion
```

File: test/unit/nektar_interface/test_particle_embed_mapping.cpp [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+#include <memory>
+#include <random>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "nektar_interface/composite_interaction/composite_interaction.hpp"
+#include "nektar_interface/particle_cell_mapping/newton_geom_interfaces.hpp"
+
+using namespace std;
```
Comment (JamesEdgeley) [inline test/unit/nektar_interface/test_particle_embed_mapping.cpp:20]:
```suggestion
```

File: test/unit/nektar_interface/test_particle_mapping.cpp [hunk 4]
```diff
@@ -14,8 +17,6 @@
 #include <string>
 #include <vector>
 
-#include "nektar_interface/particle_cell_mapping/newton_geom_interfaces.hpp"
-
 using namespace std;
```
Comment (JamesEdgeley) [inline test/unit/nektar_interface/test_particle_mapping.cpp:20]:
```suggestion
```

File: test/unit/nektar_interface/test_particle_embed_mapping.cpp [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        ASSERT_TRUE(err < 1.0e-12);
+      }
+    }
+
+    // check the map from global space back to reference space
+    REAL xi_check[3];
+    n.x_inverse(g[0], g[1], g[2], xi_check, xi_check + 1, xi_check + 2);
+    for (int dx = 0; dx < 3; dx++) {
+      const REAL err_abs = abs(xi_check[dx] - xi[dx]);
+      const REAL err = std::min(err_abs, err_abs / abs(xi[dx]));
```
Comment (JamesEdgeley) [inline test/unit/nektar_interface/test_particle_embed_mapping.cpp:95]:
```suggestion
      const REAL err = std::min(err_abs, err_abs / std::abs(xi[dx]));
```

**Review by JamesEdgeley — APPROVED** [review]
Just some namespace scoping that could be changed.  Think I commented them all.

**Comment by will-saunders-ukaea** [comment]
I have updated the NP side such that this points at the dev branch rather than an I think this is ready for another review/tests running by someone else.

**Comment by JamesEdgeley** [comment]
I get some errors like this in the build phase for the dev branch of neso-particles using spack:

```
     509    In file included from /home/jedge/NESO/neso-particles/include/neso_particles/external_interfaces/common/../../compute_target.hpp:15:
     510    In file included from /home/jedge/NESO/neso-particles/include/neso_particles/containers/../loop/../sycl_typedefs.hpp:7:
     511    In file included from /home/jedge/.spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/intel-oneapi-compilers-2022.2.1-p55yrly2lxt7mhzrgsg35vdiancszqrd/compiler/2022.2.1/linux/bin-l
            lvm/../include/sycl/CL/sycl.hpp:16:
     512    In file included from /home/jedge/.spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/intel-oneapi-compilers-2022.2.1-p55yrly2lxt7mhzrgsg35vdiancszqrd/compiler/2022.2.1/linux/bin-l
            lvm/../include/sycl/CL/sycl/backend.hpp:18:
     513    In file included from /home/jedge/.spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/intel-oneapi-compilers-2022.2.1-p55yrly2lxt7mhzrgsg35vdiancszqrd/compiler/2022.2.1/linux/bin-l
            lvm/../include/sycl/CL/sycl/detail/backend_traits_opencl.hpp:26:
     514    In file included from /home/jedge/.spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/intel-oneapi-compilers-2022.2.1-p55yrly2lxt7mhzrgsg35vdiancszqrd/compiler/2022.2.1/linux/bin-l
            lvm/../include/sycl/CL/sycl/queue.hpp:20:
  >> 515    /home/jedge/.spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/intel-oneapi-compilers-2022.2.1-p55yrly2lxt7mhzrgsg35vdiancszqrd/compiler/2022.2.1/linux/bin-llvm/../include/sycl/CL
            /sycl/handler.hpp:1242:50: error: '__usmfill<RemotePoint>' is invalid; kernel name should be forward declarable at namespace scope
     516        kernel_parallel_for<KernelName, ElementType>(KernelFunc);
```

Am I using a compatible compiler?

**Comment by JamesEdgeley** [comment]
Everything working now

**Pull request merged**